src/main/scala/calc/Compiler.scala

```diff
@@ -66,17 +80,278 @@ object Compiler {
+    }
+  }
+
+  private def captureRefs(captures: List[(String, TreeType)], scope: Scope)
+                         (implicit pos: Position): List[irt.VarRef] = {
+    captures map { p =>
+      irt.VarRef(irt.Ident(p._1))(p._2)
+    }
+  }
+
+  /** All the capture params and their corresponding VarRefs for the closure.
+    * It returns an awkward tuple to avoid finding the free variables 2 times. */
+  private def captures(closure: ClosureT, scope: Scope)
+                      (implicit pos: Position): (List[irt.ParamDef], List[irt.VarRef]) = {
+    val freeVs     = freeVars(closure)
+    val isFree     = (p: (String, _)) => freeVs.contains(p._1)
+    val cs         = scope.toList.filter(isFree)
+    val cParamDefs = captureParams(cs, scope)
+    val cVarRefs   = captureRefs(cs, scope)
+
+    (cParamDefs, cVarRefs)
+  }
+
+  /** Mangles the name according to the type. */
+  private def mangleName(name: String, tp: TreeType): String =
+    name ++ "__" ++ tp.typeSuff
+
+  /** Transforms the param list into param list for closure.
+    * Also, mangles the parameter names. */
+  private def paramsFromList(params: List[TypedTree])
+                            (implicit pos: Position): List[irt.ParamDef] = {
+    params map {
+      case IdentT(name, tp) =>
+        irt.ParamDef(
+          irt.Ident(mangleName(name, tp)),
+          irtpe.AnyType, // all params have to be AnyType
+          mutable = false, rest = false)
+
       case _ =>
-        throw new Exception(
-            s"Cannot yet compile a tree of class ${tree.getClass}")
+        throw new Exception(s"Non-identifier in argument position in function declaration.")
+    }
+  }
+
+  /** Produces the list of local variable declarations */
+  private def localParamVars(params: List[IdentT])
+                            (implicit pos: Position): irt.Tree = {
+    def localVar(param: IdentT): irt.VarDef = param match {
+      case IdentT(name, tp) =>
+        val typeCode = tp.typeSuff.charAt(0)
+        val nameM    = mangleName(name, tp)
+      val varRef   = irt.Unbox(irt.VarRef(irt.Ident(nameM))(irtpe.AnyType), typeCode)
+        val newIdent = irt.Ident(name)
+        irt.VarDef(newIdent, tp, mutable = false, varRef)
+    }
+
+    irt.Block(params.map(localVar))
+  }
+
+  /** Updates the scope with closure's parameters. */
+  private def updateScopeWithParams(params: List[TypedTree], scope: Scope): Scope = {
+    val paramTuples = params map {
+      case IdentT(name, tp) => (name, tp)
+      case _               => throw new Exception("Unexpected form of parameter.")
+    }
+
+    val paramMap = paramTuples.toMap
+
+    scope ++ paramMap  // in case of duplicates, m1 ++ m2 discards the entries from the first map
+  }
+
+  // ------------------ LIBRARY FUNCTIONS COMPILATION --------------------------
+
+  /** Returns names of all the library functions used in the program. */
+  private def libFunsUsed(tree: TypedTree): Set[String] = tree match {
```

Comment:
Consider reusing `freeVars` in this method. You should be able to write this method as a one-liner with `freeVars`.